flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java

```diff
@@ -219,6 +230,71 @@ public KubernetesWatch watchPodsAndDoCallback(
 				.watch(new KubernetesPodsWatcher(podCallbackHandler)));
 	}
 
+	@Override
+	public CompletableFuture<Void> createConfigMap(KubernetesConfigMap configMap) {
+		final String configMapName = configMap.getName();
+		return CompletableFuture.runAsync(
+			() -> this.internalClient.configMaps().inNamespace(namespace).create(configMap.getInternalResource()),
+			kubeClientExecutorService)
+			.whenComplete((ignored, throwable) -> {
+				if (throwable != null) {
+					throw new FlinkRuntimeException("Failed to create ConfigMap " + configMapName, throwable);
+				}
+			});
+	}
+
+	@Override
+	public Optional<KubernetesConfigMap> getConfigMap(String name) {
+		final ConfigMap configMap = this.internalClient.configMaps().inNamespace(namespace).withName(name).get();
+		return configMap == null ? Optional.empty() : Optional.of(new KubernetesConfigMap(configMap));
+	}
+
+	@Override
+	public CompletableFuture<Boolean> checkAndUpdateConfigMap(
+			String configMapName,
+			FunctionWithException<KubernetesConfigMap, Optional<KubernetesConfigMap>, ?> function) {
+		return FutureUtils.retry(
+			() -> CompletableFuture.supplyAsync(
+				() -> getConfigMap(configMapName)
+					.map(FunctionUtils.uncheckedFunction(configMap -> {
+						final boolean updated = function.apply(configMap).map(
+							updatedConfigMap -> {
+								this.internalClient.configMaps()
+									.inNamespace(namespace)
+									.createOrReplace(updatedConfigMap.getInternalResource());
+								return true;
+							}).orElse(false);
+						if (!updated) {
+							LOG.warn("Trying to update ConfigMap {} to {} without checking pass, ignoring.",
+								configMap.getName(), configMap.getData());
+						}
+						return updated;
+					}))
+					.orElseThrow(
+						() -> new FlinkRuntimeException("ConfigMap " + configMapName + " not exists.")),
```

Comment:
If we agree to not handle the externally deletion/update, then we do not need to retry if the ConfigMap not exists.

Reply:
I think this is independent of this question because `checkAndUpdateConfigMap` should never create a non-existing config map according to the code. If `getConfigMap()` returns `Optional.empty()` then we will throw the `FlinkRuntimeException`, retry and end up again receiving an `Optional.empty()`.

Reply:
I think this method is also lacking test coverage for exactly a case where `getConfigMap` returns `Optional.empty` or where

```
this.internalClient.configMaps()
									.inNamespace(namespace)
									.createOrReplace(updatedConfigMap.getInternalResource());
```

fails.

Reply:
What I mean is the ConfigMap could be created in the `KubernetesLeaderElectionService#Watcher`. So even the first we get a `Optional.empty()`, we could a get correct ConfigMap by retrying.

I will add a two tests here.
* ConfigMap always does not exists and retry failed
* ConfigMap exists at the very begging and retry successfully

Reply:
Ah ok, now I understand the problem. Thanks for the clarification.